### ISSUE TITLE
`upload-recording` command is broken since Connectionless API changes

### ISSUE THREAD

**Issue by andrewazores** [rh-jmc-team/container-jfr#174]
See discussion in #166 comments.

The `upload-recording` command is currently broken since the Connectionless API changes. The command attempts to open a connection to the remote target JVM and get a stream of the specified recording from this target, and then redirect that stream to the jfr-datasource endpoint. However, the target JVM connection is closed before the inner stream is opened for reading to stream to the datasource, so the POST to the datasource fails.

**Comment by andrewazores** [comment]
Here is the relevant broken command implementation. `doPost` is responsible for taking the stream returned by `getBestRecordingForName` and transferring its contents into the datasource POST request. `getBestRecordingForName` attempts to provide the stream using `TargetConnectionManager#executeConnectedTask`, but this method closes the target connection after the provided function has been executed - so by the time the call stack returns to `doPost`, the target connection is closed.

`WebServer#getRecordingDescriptor` implements essentially the same behaviour, but handles opening and closing the target connection with the proper lifecycle that would allow data to be transferred out of the target before closing the connection.

```
private ResponseMessage doPost(String targetId, String recordingName, String uploadUrl)
        throws Exception {
    Optional<InputStream> recording = getBestRecordingForName(targetId, recordingName);
    if (!recording.isPresent()) {
        throw new RecordingNotFoundException(targetId, recordingName);
    }

    InputStream stream = recording.get();
    HttpPost post = new HttpPost(uploadUrl);
    post.setEntity(
            MultipartEntityBuilder.create()
                    .addBinaryBody(
                            "file", stream, ContentType.APPLICATION_OCTET_STREAM, recordingName)
                    .build());

    try (CloseableHttpClient httpClient = httpClientProvider.get();
            CloseableHttpResponse response = httpClient.execute(post);
            stream) {
        return new ResponseMessage(
                response.getStatusLine(), EntityUtils.toString(response.getEntity()));
    }
}

Optional<InputStream> getBestRecordingForName(String targetId, String recordingName)
        throws Exception {
    Optional<IRecordingDescriptor> currentRecording =
            getDescriptorByName(targetId, recordingName);
    if (currentRecording.isPresent()) {
        return targetConnectionManager.executeConnectedTask(
                targetId,
                connection -> {
                    return Optional.of(
                            connection.getService().openStream(currentRecording.get(), false));
                });
    }

    Path archivedRecording = recordingsPath.resolve(recordingName);
    if (fs.isRegularFile(archivedRecording) && fs.isReadable(archivedRecording)) {
        return Optional.of(new BufferedInputStream(fs.newInputStream(archivedRecording)));
    }

    return Optional.empty();
}
```

**Comment by andrewazores** [comment]
I think `upload-recording` probably still works for archived recordings. From a reading, it looks to only be broken for active, in-memory recordings.

**Comment by vic-ma** [comment]
I've got the basic fix working, but I'm wondering how I should pass the connection back to `doPost` so that it can be closed later. Should I use `DownloadDescriptor` just like `getRecordingDescriptor`? Or is the purpose of that different from what I'm doing or is only for net related stuff, and I should make my own class?

**Comment by andrewazores** [comment]
Nice. I think doing something similar to the `DownloadDescriptor` probably makes sense - just some kind of struct that contains both the recording stream and the target connection, so that once the post is done both can be closed cleanly. You could really just use an Apache Commons `Pair` for this too, I think - no need to even write your own data type.

**Comment by vic-ma** [comment]
Is there a difference between the tests [`shouldReadFromDiskIfNotInTarget()`](https://github.com/rh-jmc-team/container-jfr/blob/master/src/test/java/com/redhat/rhjmc/containerjfr/commands/internal/UploadRecordingCommandTest.java#L176) and [`shouldFallThroughToDiskIfNotInMemory()`](https://github.com/rh-jmc-team/container-jfr/blob/master/src/test/java/com/redhat/rhjmc/containerjfr/commands/internal/UploadRecordingCommandTest.java#L200)?

Maybe I'm missing something, but it looks like they're duplicates of each other; only the ordering of a few Mockito calls is switched.

**Comment by andrewazores** [comment]
You're right, they look like they're duplicates. Might have been a rebase issue or just a bad copy-paste, not sure. Since you're already in there updating those tests, please go ahead and remove one of those two duplicates (in its own commit preferably).
